### PR TITLE
#707 Automatic generation of iterable method if element mapping metho…

### PR DESCRIPTION
**Pull request by sjaakd**
…d(s) are qualified

So, what about maps..

I can imagine that you have `@Mapping( target="...", qualifiedBy=Foo.class)` and you know the proper element mapping to select. But you can't do that for maps. Here you need a `keyQualifiedBy` and a `valueQualifiedBy`.

### CODE REVIEW

**Comment by agudian** [comment]
> But you can't do that for maps. Here you need a keyQualifiedBy and a valueQualifiedBy.

I think that's okay for now if we support it for iterables for convenience and don't yet do it for maps: you can still very easily provide your own implementation and delegate to the right key and value mapper methods. Plus, I can't imagine that such map mappings are required that often, especially when comparing them to iterables.

File: processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/CityDto.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    public int getPopulation() {
+        return population;
+    }
+
+    public void setPopulation(int population) {
+        this.population = population;
+    }
+
+
+
```
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/CityDto.java]: new lines... :)

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 2]
```diff
@@ -601,67 +606,75 @@ private String getSourcePresenceCheckerRef() {
...
+                    methodRef = new ForgedMethod( existingName, methodRef );
+                }
+                assignment = AssignmentFactory.createMethodReference( methodRef, null, targetType );
+                assignment.setAssignment( AssignmentFactory.createDirect( sourceReference ) );
             }
+
             return assignment;
         }
 
+
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: double emtpy line...

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 3]
```diff
@@ -250,7 +250,12 @@ public PropertyMapping build() {
 
             // No mapping found. Try to forge a mapping
             if ( assignment == null ) {
-                assignment = forgeMapOrIterableMapping( sourceType, targetType, sourceRefStr, method.getExecutable() );
+                if ( (sourceType.isCollectionType() || sourceType.isArrayType()) && (targetType.isIterableType()) ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: perhaps remove the extra parenthesis around `targetType.isIterableType()`?

File: processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/RiverEntity.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+    public int getLength() {
+        return length;
+    }
+
+    public void setLength(int length) {
+        this.length = length;
+    }
+
+
```
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/RiverEntity.java]: empty lines...

File: processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * @author Sjaak Derksen
+ */
+@Mapper
+public abstract class TopologyMapper {
+
+    public static final TopologyMapper INSTANCE = Mappers.getMapper( TopologyMapper.class );
+
+    @Mapping( target = "topologyFeatures", qualifiedBy = Rivers.class )
+    public abstract TopologyEntity mapTopology(TopologyDto dto);
```
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java]: Could you maybe add a second iterable in the source to demonstrate that once a forged iterable method was created for a qualified element mapping, that the same forged method won't be used for another non-`qualifiedBy` iterable and that a new forged iterable method is created for that one instead?

**Comment by agudian** [comment]
@sjaakd, as always you leave me speachless how fast you are with fixing these kinds of issues. :rocket:

File: processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java (hunk 5 above)
Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java]: > iterable and that a new forged iterable method is created for that one instead?

Hmmm.. Just did that.. Doesn't work... 😢  Need to think about what makes an `IterableMethod` equal

Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java]: Perhaps somehow adding the qualifier to the equality-check?

Comment (sjaakd) [inline processor/src/test/java/org/mapstruct/ap/test/selection/qualifier/iterable/TopologyMapper.java]: or something along that line 👍 .. Lets make it a bit more generic.. hold on. I'll push the changes soon.

**Comment by sjaakd** [comment]
@agudian .. have a look at my latest commit. I'll squash and commit. If your ok.

**Comment by sjaakd** [comment]
> @sjaakd, as always you leave me speachless how fast you are with fixing these kinds of issues. 🚀

Thanks, of course :smile:

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java [hunk 6]
```diff
@@ -64,5 +64,48 @@ public TypeMirror getResultType() {
...
+    private boolean equals(Object object1, Object object2) {
+        if ( object1 == null ) {
+            return (object2 == null);
+        }
+        else {
+            return object1.equals( object2 );
+        }
+    }
 }
+
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/SelectionParameters.java]: empty lines

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 7]
```diff
@@ -674,8 +674,7 @@ private Assignment forgeMapMapping(Type sourceType, Type targetType, String sour
             return assignment;
         }
 
-
-        private String getName(Type sourceType, Type targetType) {
+        private String getName(Type sourceType, Type targetType ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: extra space before the closing parenthesis

File: processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java [hunk 8]
```diff
@@ -305,6 +310,15 @@ public boolean equals(Object obj) {
             }
         }
 
+        if ( this.selectionParameters != null ) {
+            if ( !this.selectionParameters.equals( other.selectionParameters ) ) {
+                return false;
+            }
+        }
+        else if ( other.selectionParameters == null ) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java]: I think this needs to be a `!=`... ?

**Comment by agudian** [comment]
Looks good! Check out my last comments and go ahead with squashing and pushing...  :+1:

File: processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java (hunk 8 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/IterableMappingMethod.java]: yep

**Pull request closed without merging**